lib/srv/db/postgres/users.go

```diff
@@ -81,8 +87,201 @@ func (e *Engine) ActivateUser(ctx context.Context, sessionCtx *common.Session) e
+		return nil
+	}
+
+	if len(sessionCtx.DatabaseRoles) > 0 {
+		e.Log.Errorf("Cannot apply fine-grained permissions: non-empty list of database roles (%v).", sessionCtx.DatabaseRoles)
+		return trace.BadParameter("fine-grained database permissions and database roles are mutually exclusive, yet both were provided.")
+	}
+
+	rules, err := e.AuthClient.GetDatabaseObjectImportRules(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	conn, err := e.connectAsAdmin(ctx, sessionCtx, false)
+	if err != nil {
+		e.Log.WithError(err).Error("Failed to connect to the database.")
+		return trace.Wrap(err)
+	}
+	defer conn.Close(ctx)
+
+	objsFetched, err := fetchDatabaseObjects(ctx, sessionCtx, conn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	counts, _ := permissions.CountObjectKinds(objsFetched)
+	e.Log.Infof("Fetched %v objects from the database (%v).", len(objsFetched), counts)
+
+	objsTagged := permissions.ApplyDatabaseObjectImportRules(rules, sessionCtx.Database, objsFetched)
+	counts, _ = permissions.CountObjectKinds(objsTagged)
+	e.Log.Infof("Tagged %v database objects (%v).", len(objsTagged), counts)
+
+	permissionSet, err := permissions.CalculatePermissions(sessionCtx.Checker, sessionCtx.Database, objsTagged)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	summary, eventData := permissions.SummarizePermissions(permissionSet)
+	e.Log.Infof("Calculated database permissions for user %q: %v.", sessionCtx.DatabaseUser, summary)
+	e.auditUserPermissions(sessionCtx, eventData)
+
+	perms, err := convertPermissions(permissionSet)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// teleport_remove_permissions and teleport_update_permissions are created in pg_temp table of the session database.
+	// teleport_remove_permissions gets called by teleport_update_permissions as needed.
+	_, err = conn.Exec(ctx, removePermissionsProc)
+	if err != nil {
+		e.Log.WithError(err).Errorf("Creating temporary stored procedure %q failed.", removePermissionsProcName)
+		return trace.Wrap(err)
+	}
+
+	_, err = conn.Exec(ctx, updatePermissionsProc)
+	if err != nil {
+		e.Log.WithError(err).Errorf("Creating temporary stored procedure %q failed.", updatePermissionsProcName)
+		return trace.Wrap(err)
+	}
+
+	_, err = conn.Exec(ctx, updatePermissionsQuery, sessionCtx.DatabaseUser, perms)
+	if err != nil {
+		var pgErr *pq.Error
+		if errors.As(err, &pgErr) {
+			if pgErr.Code == common.SQLStatePermissionsChanged {
+				e.Log.WithError(err).Errorf("Permissions have changed, rejecting connection for user %q.", sessionCtx.DatabaseUser)
+			}
+		} else {
+			e.Log.WithError(err).Errorf("Failed to update permissions for user %q.", sessionCtx.DatabaseUser)
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (e *Engine) removePermissions(ctx context.Context, sessionCtx *common.Session) error {
+	if !e.granularPermissionsEnabled(sessionCtx) {
+		e.Log.Infof("Granular database permissions not enabled, skipping removal step for user %q.", sessionCtx.DatabaseUser)
```

Comment:
nitpick: per [0154 logging guide line](https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md) `WithField` should be used when possible

> Embrace structured logging. Prefer using static messages and attributes instead of relying on fmt.Sprintf style formatting. Enrich log messages with additional attributes to provide additional context.

Reply:
Done.